probe/proc/cache.go

```diff
@@ -0,0 +1,123 @@
+package proc
+
+import (
+	"bytes"
+	"sync"
+	"time"
+
+	lru "github.com/hashicorp/golang-lru"
+)
+
+const (
+	filesCacheLen        = 1024
+	filesCacheExpiration = 60 * time.Second
+)
+
+type filesCacheEntry struct {
+	sync.RWMutex
+	file     File
+	contents *bytes.Buffer
+	creation time.Time
+}
+
+func (fce *filesCacheEntry) Pinned() bool { return fce.contents != nil }
+func (fce *filesCacheEntry) ReadInto(buf *bytes.Buffer) error {
+	fce.Lock()
+	defer fce.Unlock()
+
+	if fce.Pinned() { // use the saved contents
+		buf.Write(fce.contents.Bytes())
+		return nil
+	}
+	return fce.file.ReadInto(buf)
+}
+
+// A cache for files handles
+type filesCache struct {
+	*lru.Cache
+	proc Dir
```

Comment:
I think we don't really need that. `Dir` is only used here for `Open()`ing files, and calling `Open()` from several goroutines for the same file is probably safe. So the thing that really must be goroutines-safe is the `File` returned by `Dir.Open()`, and the cache protects accesses to that `FIle` with the `RWLock` entries hold...